plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,401 @@
+{
+	/* PSCI v0.1 and v0.2 both support cpu_on. the parameter @cpu after
+	 * @entry will be passed as x0 for entry
+	 */
+	return smcc_psci_hvc_call(PSCI_FNID_CPU_ON, (__u64) cpu->id,
+					(__u64) entry, (__u64) cpu);
+}
+
+int ukplat_lcpu_start(__lcpuid lcpuid[], void *sp[],
+				ukplat_lcpu_entry_t entry[], unsigned int num)
+{
+	__uptr lcpu_start_addr = ukplat_virt_to_phys(_lcpu_start);
+	__u32 this_cpu_id = ukplat_lcpu_id();
+	__lcpuid cpu_idx;
+	__u32 i, n = 0;
+	int err;
+
+	UK_ASSERT(sp);
+
+	if (num == 0) {
+		UK_ASSERT(!lcpuid);
+		num = smp_numcores;
+	}
+
+
+	/* register the lcpu_run and lcpu_wakeup interrupt handlers */
+	err = ukplat_irq_register(UKPLAT_ARM64_IPI_RUN_IRQ,
+					arm64_lcpu_ipi_run_handler, NULL);
+	if (err != 0) {
+		uk_pr_crit("Interrupt handler for IRQ %d failed\n",
+			UKPLAT_ARM64_IPI_RUN_IRQ);
+		return err;
+	}
+
+	/* start the secondary cores */
+	for (i = 0; i < num; i++) {
+		cpu_idx = (lcpuid) ? lcpuid[i] : (__lcpuid) i;
+
+		if (cpus[cpu_idx].id == this_cpu_id)
+			continue;
+
+		if (cpu_idx >= smp_numcores)
+			continue;
+
+		if (cpus[cpu_idx].state != LCPU_STATE_OFFLINE)
+			uk_pr_warn("Core %d already starting, \
+				moving on.\n", cpu_idx);
+
+		if (ukarch_spin_trylock(&cpus[cpu_idx].lock)) {
+			cpus[cpu_idx].entry = entry[i] ? entry[i] : _lcpu_entry_default;
+			cpus[cpu_idx].stackp = sp[i];
+			cpus[cpu_idx].state = LCPU_STATE_STARTING;
+			UK_INIT_LIST_HEAD(&cpus[cpu_idx].fnlist);
+
+			barrier();
+
+			err = psci_cpu_on(&(cpus[cpu_idx]), lcpu_start_addr);
+			if (err != PSCI_RET_SUCCESS) {
+				uk_pr_err("Failed to start CPU %u (%u): %d\n",
+					cpu_idx, cpus[cpu_idx].id, err);
+				cpus[cpu_idx].state = LCPU_STATE_OFFLINE;
+				ukarch_spin_unlock(&(cpus[cpu_idx].lock));
+				continue;
+			}
+
+			uk_pr_info("Started core (%u) successfully\n",
+				cpus[cpu_idx].id);
+			ukarch_spin_unlock(&cpus[cpu_idx].lock);
+			n++;
+		}
+	}
+
+	return n;
+}
+
+void __noreturn _lcpu_entry_default(void)
+{
+	intctrl_init();
+	ukplat_lcpu_enable_irq();
+
+	/* Enter halt loop. Besides interrupts in general, the halt can be
+	 * interrupted by calls to ukplat_lcpu_run().
+	 */
+	while (1) {
+		cpus[ukplat_lcpu_idx()].state = LCPU_STATE_IDLE;
+		halt();
+	}
+}
+
+int ukplat_lcpu_run(__lcpuid lcpuid[], unsigned int num,
+			struct ukplat_lcpu_func *fn, int flags)
+{
+	__u32 i;
+	__u32 this_cpu_id = ukplat_lcpu_id();
+	__lcpuid cpu_idx;
+
+	UK_ASSERT(fn);
+
+	if (num == 0) {
+		UK_ASSERT(!lcpuid);
+		num = smp_numcores;
+	}
+	if (num > smp_numcores) {
+		uk_pr_warn("More cores are selected, but not available\n");
+		num = smp_numcores;
+	}
+
+	for (i = 0; i < num; i++) {
+		cpu_idx = (lcpuid) ? lcpuid[i] : (__lcpuid) i;
+
+		if (cpus[cpu_idx].id == this_cpu_id)
+			continue;
+
+		ukarch_spin_lock(&cpus[cpu_idx].lock);
+		uk_list_add(&fn->lentry, &cpus[cpu_idx].fnlist);
+		intctrl_send_ipi(UKPLAT_ARM64_IPI_RUN_IRQ, cpus[cpu_idx].id);
+	}
+
+	return 0;
+}
+
+int __attribute__((optimize("O0")))
```

Comment:
I used it in order to make sure the compiler won't optimize the loop. However, I realized that since the state field of the `arm_cpu` structure is marked as volatile, we can remove it.